### PR TITLE
Move the setup of external share mounts to a mountprovider

### LINKED ISSUE #15116: Remote share mount point cannot be reshared with local users
I'm just installed two OC8.0.2 and I use a public shared folder to sync data between each OC.
In the "master" OC, I have created the folder, shared it by a public share, and set permission for the local users.
In the "slave" OC, I have "connected" the remote shared folder and set permission for the local users with the admin account.

When an user of the slave OC log in the web gui of the slave OC, he can access to the remote share folder and work with it. But this public folder is not visible by webdav and OC client (Windows).

https://forum.owncloud.org/viewtopic.php?uid=25511&f=31&t=27290&start=0

### PR DESCRIPTION
**Pull request by icewind1991**
Fixes #15116

Makes it possible to reshare an incoming external share

cc @PVince81 @DeepDiver1975 @MorrisJobke

### CODE REVIEW

File: apps/files_sharing/lib/external/mountprovider.php [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+	 * @param callable $managerProvider due to setup order we need a callable that return the manager instead of the manager itself
+	 */
+	public function __construct(IDBConnection $connection, callable $managerProvider) {
+		$this->connection = $connection;
+		$this->managerProvider = $managerProvider;
+	}
+
+	public function getMount(IUser $user, $data, IStorageFactory $storageFactory) {
+		$data['manager'] = $this;
+		$mountPoint = '/' . $user->getUID() . '/files' . $data['mountpoint'];
```
Comment (PVince81) [inline apps/files_sharing/lib/external/mountprovider.php]: don't assume that `$data['mountpoint']` starts with a slash. In the past we already got enough nasty surprises.
Use `ltrim($data['mountpoint'], '/') . '/'`

Comment (icewind1991) [inline apps/files_sharing/lib/external/mountprovider.php]: Done

**Comment by PVince81** [comment]
The code looks good but it doesn't seem to completely solve the reshare issue.

I tried the steps here https://github.com/owncloud/core/issues/15116#issuecomment-95592730 but on a single instance.

"buser1" can see the folder "fold" and enter it, but cannot upload.
Then I went into the folder as "admin" (the recipient) and was able to upload stuff into "fold".
Then I went back with "buser1" and not I cannot enter the folder at all:
```
{"reqId":"DSfKoz2ySol1uxgKtLUK","remoteAddr":"127.0.0.1","app":"files","message":"Exception: {\"Exception\":\"Sabre\\\\HTTP\\\\ClientHttpException\",\"Message\":\"Not Found\",\"Code\":404,\"Trace\":\"
#0 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Client.php(233): Sabre\\\\HTTP\\\\Client->send(Object(Sabre\\\\HTTP\\\\Request))\\n
#1 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/files\\\/storage\\\/dav.php(265): Sabre\\\\DAV\\\\Client->propFind('fold', Array)\\n
#2 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/files\\\/storage\\\/dav.php(679): OC\\\\Files\\\\Storage\\\\DAV->propfind('fold')\\n
#3 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/files\\\/storage\\\/wrapper\\\/wrapper.php(446): OC\\\\Files\\\\Storage\\\\DAV->getETag('fold')\\n
#4 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/files\\\/storage\\\/wrapper\\\/availability.php(412): OC\\\\Files\\\\Storage\\\\Wrapper\\\\Wrapper->getETag('fold')\\n
#5 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/files\\\/storage\\\/wrapper\\\/wrapper.php(446): OC\\\\Files\\\\Storage\\\\Wrapper\\\\Availability->getETag('fold')\\n
#6 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/files_sharing\\\/lib\\\/sharedstorage.php(591): OC\\\\Files\\\\Storage\\\\Wrapper\\\\Wrapper->getETag('fold')\\n
#7 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/files\\\/storage\\\/common.php(624): OC\\\\Files\\\\Storage\\\\Shared->getETag('')\\n
#8 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/files\\\/storage\\\/wrapper\\\/wrapper.php(561): OC\\\\Files\\\\Storage\\\\Common->getMetaData('')\\n
#9 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/files\\\/storage\\\/wrapper\\\/availability.php(456): OC\\\\Files\\\\Storage\\\\Wrapper\\\\Wrapper->getMetaData('')\\n
#10 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/files\\\/storage\\\/wrapper\\\/wrapper.php(561): OC\\\\Files\\\\Storage\\\\Wrapper\\\\Availability->getMetaData('')\\n
#11 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/files\\\/cache\\\/scanner.php(113): OC\\\\Files\\\\Storage\\\\Wrapper\\\\Wrapper->getMetaData('')\\n
#12 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/files_sharing\\\/lib\\\/scanner.php(39): OC\\\\Files\\\\Cache\\\\Scanner->getData('')\\n
#13 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/files\\\/cache\\\/scanner.php(141): OC\\\\Files\\\\Cache\\\\SharedScanner->getData('')\\n
#14 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/files\\\/cache\\\/scanner.php(265): OC\\\\Files\\\\Cache\\\\Scanner->scanFile('', 3, -1, NULL, true)\\n
#15 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/files\\\/cache\\\/watcher.php(91): OC\\\\Files\\\\Cache\\\\Scanner->scan('', false)\\n
#16 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/files_sharing\\\/lib\\\/watcher.php(41): OC\\\\Files\\\\Cache\\\\Watcher->checkUpdate('', Array)\\n
#17 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/files\\\/view.php(1203): OC\\\\Files\\\\Cache\\\\Shared_Watcher->checkUpdate('', Array)\\n
#18 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/files\\\/filesystem.php(861): OC\\\\Files\\\\View->getFileInfo('\\\/fold', true)\\n
#19 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/files\\\/ajax\\\/list.php(35): OC\\\\Files\\\\Filesystem::getFileInfo('\\\/fold')\\n
#20 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/route\\\/route.php(154) : runtime-created function(1): require_once('\\\/srv\\\/www\\\/htdocs...')\\n
#21 [internal function]: __lambda_func(Array)\\n
#22 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/route\\\/router.php(281): call_user_func('\\\\x00lambda_609', Array)\\n
#23 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/base.php(829): OC\\\\Route\\\\Router->match('\\\/apps\\\/files\\\/aja...')\\n
#24 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/index.php(40): OC::handleRequest()\\n
#25 {main}\",\"File\":\"\\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/3rdparty\\\/sabre\\\/http\\\/lib\\\/Client.php\",\"Line\":154}","level":3,"time":"2015-10-05T15:13:35+00:00","method":"GET","url":"\/owncloud\/index.php\/apps\/files\/ajax\/list.php?dir=%2Ffold&sort=name&sortdirection=asc"}
```

How did you test ?

**Comment by icewind1991** [comment]
> How did you test ?

Only for single files, will look into the folder issues

**Comment by icewind1991** [comment]
Fixed working with re-shared folders

**Comment by icewind1991** [comment]
@MorrisJobke @DeepDiver1975 @PVince81 please review

**Comment by PVince81** [comment]
Works now with folders :+1:

**Comment by icewind1991** [comment]
@MorrisJobke @DeepDiver1975 @schiesbn can you review this?

**Comment by MorrisJobke** [comment]
Tested and works :+1:

Before it was simply not possible to see the share - now you can browse it

**Comment by MorrisJobke** [comment]
@icewind1991 @DeepDiver1975 Merge this please

**Pull request merged**